Kernel/Arch/aarch64/PageDirectory.cpp

```diff
@@ -5,35 +5,120 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include <AK/Singleton.h>
+#include <Kernel/Arch/CPU.h>
+#include <Kernel/Arch/PageDirectory.h>
 #include <Kernel/Arch/aarch64/ASM_wrapper.h>
-#include <Kernel/Memory/PageDirectory.h>
+#include <Kernel/InterruptDisabler.h>
+#include <Kernel/Memory/MemoryManager.h>
+#include <Kernel/Prekernel/Prekernel.h>
+#include <Kernel/Process.h>
+#include <Kernel/Random.h>
+#include <Kernel/Sections.h>
+#include <Kernel/Thread.h>
 
 namespace Kernel::Memory {
 
-void PageDirectory::register_page_directory(PageDirectory*)
+struct TTBR0Map {
+    SpinlockProtected<IntrusiveRedBlackTree<&PageDirectory::m_tree_node>, LockRank::None> map {};
+};
+
+static Singleton<TTBR0Map> s_ttbr0_map;
+
+void PageDirectory::register_page_directory(PageDirectory* directory)
 {
-    dbgln("FIXME: PageDirectory: Actually implement registering a page directory!");
+    s_ttbr0_map->map.with([&](auto& map) {
+        map.insert(directory->ttbr0(), *directory);
+    });
 }
 
-void PageDirectory::deregister_page_directory(PageDirectory*)
+void PageDirectory::deregister_page_directory(PageDirectory* directory)
 {
-    TODO_AARCH64();
+    s_ttbr0_map->map.with([&](auto& map) {
+        map.remove(directory->ttbr0());
+    });
 }
 
 LockRefPtr<PageDirectory> PageDirectory::find_current()
 {
-    TODO_AARCH64();
-    return nullptr;
+    return s_ttbr0_map->map.with([&](auto& map) {
+        return map.find(Aarch64::Asm::get_ttbr0_el1());
+    });
+}
+
+void activate_kernel_page_directory(PageDirectory const& page_directory)
+{
+    Aarch64::Asm::set_ttbr0_el1(page_directory.ttbr0());
+}
+
+void activate_page_directory(PageDirectory const& page_directory, Thread* current_thread)
+{
+    current_thread->regs().ttbr0_el1 = page_directory.ttbr0();
+    Aarch64::Asm::set_ttbr0_el1(page_directory.ttbr0());
+}
+
+UNMAP_AFTER_INIT NonnullLockRefPtr<PageDirectory> PageDirectory::must_create_kernel_page_directory()
+{
+    return adopt_lock_ref_if_nonnull(new (nothrow) PageDirectory).release_nonnull();
 }
 
-void activate_kernel_page_directory(PageDirectory const&)
+ErrorOr<NonnullLockRefPtr<PageDirectory>> PageDirectory::try_create_for_userspace()
+{
+    auto directory = TRY(adopt_nonnull_lock_ref_or_enomem(new (nothrow) PageDirectory));
+
+    directory->m_root_table = TRY(MM.allocate_physical_page());
+
+    directory->m_directory_table = TRY(MM.allocate_physical_page());
+    auto kernel_pd_index = (kernel_mapping_base >> 30) & 0x1ffu;
```

Comment:
kind of magic (numbers)? would be nice to link that to something (or attach a comment?).

Reply:
You are totally right that this code can have some more abstraction to make it easier to understand, but that can be part of a different PR :) (Maybe by you? :yakbait: ;p)

Reply:
hahah! maybe! yak or no — definitely will help me to understand stuff!